vernac/vernacentries.ml

```diff
@@ -2749,7 +2749,10 @@ and vernac_load ~verbosely ~st fname =
 
 and interp_control ?proof ~st v = match v with
   | { v=VernacExpr (atts, cmd) } ->
-    interp_expr ?proof ~atts ~st cmd
+    let before_univs = Global.universes () in
+    let pstack = interp_expr ?proof ~atts ~st cmd in
+    if before_univs == Global.universes () then pstack
+    else Option.map Proof_global.(modify_current_pstate update_global_env) pstack
```

Comment:
Actually this is not correct, eg if you load
~~~coq
Definition idw (A : Type) := A.
Set Nested Proofs Allowed.

Lemma foobar' : unit.
  Lemma hide : True. Proof.
  Lemma aux : forall A : Type, A -> unit.
  Proof. intros. pose (foo := idw A). exact tt. Show Universes. Qed.
  exact I. Qed.
  apply (@aux unit tt).
Qed.
~~~
Not sure if we need to modify all pstates (instead of just the top one) or what.
The same code run in the regular way works fine (it's using the stm update_global_env instead I guess).

Reply:
I guess we can go with this for now, I doubt anyone uses Load on something with nested proofs since it already doesn't work in master.

Reply:
Even if the PR is correct as this IMHO we should reify the universe modifications into something that can allow clients to handle such effect in a more controlled way, so for example a tactic run will return a `result` type kind of `Nothing | UpdateUnivs of universe`.

Reply:
So what are the paths new universes get created?

Reply:
On the other hand it is reasonable to forbid load from every opening a proof.

Reply:
> So what are the paths new universes get created?

Also something I'm wondering is why do we do a merge, isn't the process of creating new universes monotonous [so an addition would suffice]?

But something is strange here, imagine we have:
```
Lemma p1 : T. Proof. p1. Qed.
Lemma p2 : T. Proof. p2. Qed.
```
where both proofs do create universes, it turns out we cannot really run them in parallel, right? We must use the "global" counter and keep the corresponding locks, etc....

So it seems to me that indeed new universe creation should go into a special part of the `evar_map` so that opaque proofs do return the set of universes created in the proof and then these are reconciled; likely by an update of the universe map as to be more efficient. WDYT?

Reply:
For now what I'm planning to do with delegated proofs is to check if they are producing new universes, and if so, re-schedule them to be run in a non-delegated way.

Reply:
> For now what I'm planning to do with delegated proofs is to check if they are producing new universes, and if so, re-schedule them to be run in a non-delegated way.

No please. All proofs can generate new universes. There is a remote_counter interface (with locks and stuff). Also "re-scheduling" things is just extra complexity (wrt the already existing global remote counter with a lock).

The orthodox solution to get rid of this complexity is to make universe names depend on the constant in which they were generated, making them "global to a constant" instead of "global to the system". If it was done that way no lock would be needed, no global counter.

BTW, a while ago, if I'm not mistaken, @SkySkimmer make universes created during proof construction more hidden than before. Bet I don't recall the details.

Reply:
Having the tactic engine do IPC is just too heavy to me; anyways that's a valid choice as a client implementor, to trade off complexity for performance.

> The orthodox solution to get rid of this complexity is to make universe names depend on the constant in which they were generated, making them "global to a constant" instead of "global to the system". If it was done that way no lock would be needed, no global counter.

Even if you do so [which is what I was proposing above], you still have a question about combination of such local universes and that seems far from trivial I think.

Reply:
Delegated proofs can produce different universes and constraints compared to non-delegated proofs, including different unification failures and all that ensues.
Most scripts don't trigger such issues so in practice we just ignore the possibility when we want to run proofs in parallel.

>So what are the paths new universes get created?

Anything using constrs + a couple commands like Universe

>BTW, a while ago, if I'm not mistaken, @SkySkimmer make universes created during proof construction more hidden than before. Bet I don't recall the details.

I don't know what you're talking about.

Reply:
> Anything using constrs + a couple commands like Universe

Yeah thanks, I wonder how frequent that is in practice; I guess I'll instrument the CI once I can push some changes.

I guess from the User Interface point of view we want to mark such proofs and dependencies as unsafe, they could be added to a queue and re-checked in the background in the proper safe order, I just wish there was a way to construct a DAG-like dependency tree instead of just the linear order we'd do today.